### ISSUE TITLE
Incorrect notes being played

### ISSUE THREAD

**Issue by nickjohnson-dev** [Tonejs/Tone.js#163]
I have been using Tone for a few months now in a project, and just now noticed some really odd behavior.

When triggering a note's attack, the right note is not always played.

I have reproduced the behavior in a simple JSBin:

http://jsbin.com/yefilahapu/edit?js,output

Clicking the button repeatedly should produce alternating octaves of C, but at least for me, every few presses, it will repeat the same note the next 1-4 times you press it.

Also, sometimes a seemingly random note is played, that is neither C3 or C4.

Am I missing something in the way I'm configuring the synth, or is this a bug?

I don't remember running into this issue before 0.7.x, but downgrading didn't seem to fix the issue.

**Comment by tambien** [comment]
hmmm. strange. I'm unable to reproduce this issue with your jsbin. I just spent about 10 minutes clicking at all different rates using Chrome, FF, and Safari. I didn't hear it once. 

I'm on a mid 2014 Macbook Pro OS X 10.10.5.

What system/OS are you using? Can you try with 'sine' instead of 'square' as the oscillator type and see if the issue still exists? that might narrow down the cause.

**Comment by nickjohnson-dev** [comment]
I'm on a late-2013 MacBook Pro with OS X 10.11.6.
I've also been using Chrome 54.0.2840.16 beta (64-bit).

I just tried it out in Safari myself, and the issue appears to be gone.

Thinking that it was just a beta bug, I got back on the stable release channel just now and tried again, but I'm still having the issue in Chrome.

**Comment by tambien** [comment]
crazy! I was on Chrome 52, but just upgraded and i'm hearing that issue now. I'm following all of the values down to where they schedule the AudioParam and everything looks accurate on the Tone.js-side of things. I'm trying to see if i can put together a minimal test to see if it's a Chrome-related issue.

**Comment by tambien** [comment]
I submitted [a bug to Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=645776). 

Please star it to boost the priority of this issue being looked into.

**Comment by nickjohnson-dev** [comment]
Done!

Thanks for taking a look at it.

**Comment by stuartpb** [comment]
Since this seems to be a bug regarding `setValueAtTime(0, ...`, would it make sense to modify the code to just set the value directly, instead of calling `setValueAtTime`?

**Comment by tambien** [comment]
@stuartpb If you're only concerned with immediate value changes, that should work fine. But it obviously won't work for any kind of precise timing. 

Maybe that's a temporary fix until this Chrome issue is resolved. Something like: if time < currentTime + someSmallLookAhead, then just set .value instead of setValueAtTime. It's better to have it change, albeit slightly imprecisely than to miss the scheduling altogether.

**Comment by tambien** [comment]
@stuartpb So i was starting to implement your suggestion which was that if the time is close to the currentTime, just set the AudioParam.value instead of scheduling a change (though according to the specification they are supposed to be the same), but i think it uncovered [another bug](https://bugs.chromium.org/p/chromium/issues/detail?id=649681) in Chrome's AudioParam which is that the value is supposed to change immediately, but instead it ramps between values... so that still doesn't seem to fix this issue in Chrome.

**Comment by tambien** [comment]
alright! i implemented @stuartpb's suggestion and i think it fixed it. 

Here's [your jsbin](http://jsbin.com/gutoretimi/1/edit?js,output) running the latest build from the dev branch. If you have a second, verify that it works for you. It seems to be working for me. I haven't heard the error once.

**Comment by stuartpb** [comment]
Why the `+ someSmallLookAhead`? Isn't `time <= currentTime` enough?

**Comment by tambien** [comment]
@stuartpb in my testing, AudioParam events that are scheduled very close to the currentTIme also fail to set sometimes with setValueAtTime in Chrome.

**Comment by esbullington** [comment]
Hey, great library!  Very new to ToneJS, but I'm on Chrome 53.02785 (64-bit, linux) and I'm running into a problem I believe may be related to this Chrome bug.  The patch on the dev branch doesn't seem to address it.

When I run this function, I first hear a single note of A4 before the C2 tone starts kicking in:

```js
function startSynth() {
  const synth = new Tone.Synth().toMaster();
  const loop = new Tone.Loop(function(time){
      synth.triggerAttackRelease("C2", "4n", time);
  }, "2n");
  loop.start(0);
  // "+0.2": Temporarily work around
  // for this Chrome bug: 
  // https://bugs.chromium.org/p/chromium/issues/detail?id=645776
  // https://github.com/Tonejs/Tone.js/issues/163
  // loop.start('+0.1');
};
// elsewhere I call startSync() and then Tone.Transport.start()
```

As you can see, if I schedule the loop just a tenth of a second later, it avoids this problem of the initial A4 note.

Could this be related to the Chrome "ramping up" issue cited above, or is this normal and my expectations are wrong?  If it's likely the Chrome bug, then unfortunately the dev branch doesn't seem to take care of the problem (I'm using `npm link` to use the library locally).

**Comment by tambien** [comment]
@esbullington Did you build the latest dev branch? The build that's currently there is still r7, so if you're importing without rebuilding (or requiring the src files individually) you might not be running the latest. I just tested it and didn't have any issue. [Here's](http://jsbin.com/wuwohiyujo/1/edit?js,output) your example running with the latest build. Let me know if that works for you.

**Comment by esbullington** [comment]
@tambien Thanks for your response.  So you're right, I had neglected to re-build the dev branch.  So when I did that, it took care of the extra note issue, sorry for the confusion.

Unfortunately, even after I re-built the dev branch, I still get a very rapid initial beep or squeak upon `Tone.Transport.start()` on both a Chromebook Asus  C100P as well as on an Intel  i5-2500K CPU @ 3.30GHz w/32GB of RAM.  Oddly, this beep didn't occur or else wasn't noticeable when the Chrome ramping up bug was active (possibly because the beep is at A4???).

This appears to perhaps be the same issue as #116, so I would be glad to take this discussion over there. 

I frequently (4 out of 5 tries) get the same beep running the JSBin example you link to above on the Chromebook running the latest Chrome (53.02785).  In fact, on the Chromebook, I even get the beep/squeak after rolling Chrome back to 52.0.2743.116.

Oddly, when testing your JSBin on the i5,  I haven't heard the sound (3-4 tries with no initial beep).  But I hear the beep when  I test the same code on my app (same except the UI is a React/Redux app).

Here's [a recording](https://drive.google.com/file/d/0BwENGyNjJsKqRk16emlWQmpwaGM/view?usp=sharing) of the pop/beep/squeak at the beginning of the loop.  This is on the Asus C100P Chromebook running Chrome  52.0.2743.116 (since I haven't yet un-rolled the Chromebook back to the latest Chrome) and the JSBin you made above for the dev branch.  It actually took me three tries to get this recording, the first two were fine.

I recognize that this issue may be a bug with Web Audio API on Chrome, and not with ToneJS.  As such, I mention them just for your information and not with any expectations of a fix.

**Comment by tambien** [comment]
@esbullington I haven't heard this specific glitch that you documented, but I find in general when i pass in a small value in the future to `Transport.start` a lot of the glitches of first time events go away. 

for example: `Tone.Transport.start("+0.1")` instead of `Tone.Transport.start()` seems to fix a lot of my glitches. I think I'm going to be updating the examples to use that pattern.

**Comment by nickjohnson-dev** [comment]
@tambien I tested the JSBin, and it works for me as well. Great work.

Do you plan to release an updated npm package with the changes in the dev branch? I'm not exactly in a rush since my app isn't exactly in production with users that depend on it working, but I was just curious.

**Comment by tambien** [comment]
I just pushed r8. It will hopefully fix many of these scheduling issues on Chrome until they release a fix for it. (fingers crossed).

**Comment by tambien** [comment]
I'm reopening this issue, but it is out of my hands until Chrome releases a fix.

**Comment by pikurasa** [comment]
We do not seem to experience the bug since we updated tone.js on Jan 11th. See https://github.com/walterbender/musicblocks/issues/295#issuecomment-275160656 and https://github.com/walterbender/musicblocks/pull/468

If this is the case, the issue was probably not due to anything on the Chromium side.

**Comment by tambien** [comment]
Glad to hear that this bug was mitigated in the latest Tone.js. I think it's probably due to some of the Tone.Clock updates that i made in which notes get scheduled further ahead of time which reduces the frequency of that error occurring, not the fact that this bug is fixed (or not present) in Chrome. 

[This fiddle](https://jsfiddle.net/yotammann/8jrudcs8/10/) doesn't use any Tone.js and still exhibits the bug. so we're still waiting on Chrome to address this issue. The changes that I made do not guarantee that the error will not happen in musicblocks, but they will hopefully greatly reduce the probability of them happening.

**Comment by tambien** [comment]
Issue is fixed in Chrome 57! Just verified myself. I will go ahead and remove [this temporary fix](10459f42817402db5c58e2acf81e0cfbfc743be9 )

**Comment by nickjohnson-dev** [comment]
That's great news! Thanks for the hard work in providing the temporary fix and following the issue.
